pyroomacoustics/bss/auxiva.py

```diff
@@ -1,131 +1,251 @@
 
-    # initialize the demixing matrices
+    W_hat = np.zeros((n_freq, n_chan, n_chan), dtype=X.dtype)
+    W = W_hat[:, :n_src, :]
+    J = W_hat[:, n_src:, :n_src]
+
+    def tensor_H(T):
+        return np.conj(T).swapaxes(1, 2)
+
+    def update_J_from_orth_const():
+        tmp = np.matmul(W, Cx)
+        J[:, :, :] = tensor_H(np.linalg.solve(tmp[:, :, :n_src], tmp[:, :, n_src:]))
+
+    # initialize A and W
     if W0 is None:
-        W = np.array([np.eye(n_chan, n_src) for f in range(n_freq)], dtype=X.dtype)
+
+        if init_eig:
+            # Initialize the demixing matrices with the principal
+            # eigenvectors of the input covariance
+            v, w = np.linalg.eig(Cx)
+            for f in range(n_freq):
+                ind = np.argsort(v[f])[-n_src:]
+                W[f, :, :] = np.conj(w[f][:, ind])
+
+        else:
+            # Or with identity
+            for f in range(n_freq):
+                W[f, :, :n_src] = np.eye(n_src)
+
     else:
-        W = W0.copy()
+        W[:, :, :] = W0
 
-    if f_contrast is None:
-        f_contrast = f_contrasts['norm']
-        f_contrast_args = [1, 1]
+    # We still need to initialize the rest of the matrix
+    if n_src < n_chan:
+        update_J_from_orth_const()
+        for f in range(n_freq):
+            W_hat[f, n_src:, n_src:] = -np.eye(n_chan - n_src)
 
-    I = np.eye(n_src,n_src)
-    Y = np.zeros((n_frames, n_freq, n_src), dtype=X.dtype)
-    r = np.zeros((n_frames, n_src))
-    G_r = np.zeros((n_frames, n_src))
+    eyes = np.tile(np.eye(n_chan, n_chan), (n_freq, 1, 1))
+    V = np.zeros((n_freq, n_chan, n_chan), dtype=X.dtype)
+    r_inv = np.zeros((n_src, n_frames))
+    r = np.zeros((n_src, n_frames))
+
+    # Things are more efficient when the frequencies are over the first axis
+    Y = np.zeros((n_freq, n_src, n_frames), dtype=X.dtype)
+    X_original = X
+    X = X.transpose([1, 2, 0]).copy()
 
     # Compute the demixed output
     def demix(Y, X, W):
-        for f in range(n_freq):
-            Y[:,f,:] = np.dot(X[:,f,:], np.conj(W[f,:,:]))
+        Y[:, :, :] = np.matmul(W, X)
 
     for epoch in range(n_iter):
 
         demix(Y, X, W)
 
         if callback is not None and epoch % 10 == 0:
+            Y_tmp = Y.transpose([2, 0, 1])
             if proj_back:
-                z = projection_back(Y, X[:,:,0])
-                callback(Y * np.conj(z[None,:,:]))
+                z = projection_back(Y_tmp, X_original[:, :, 0])
+                callback(Y_tmp * np.conj(z[None, :, :]))
             else:
-                callback(Y)
+                callback(Y_tmp)
 
         # simple loop as a start
         # shape: (n_frames, n_src)
-        r[:,:] = np.sqrt(np.sum(np.abs(Y * np.conj(Y)), axis=1))
+        if model == "laplace":
+            r[:, :] = 2.0 * np.linalg.norm(Y, axis=0)
+        elif model == "gauss":
+            r[:, :] = (np.linalg.norm(Y, axis=0) ** 2) / n_freq
```

Comment:
Would it be more clean to define `gauss` and `laplace` as separate functions and call them here? Or creating a dictionary from where we select the desired function based on `model`.

Reply:
Because the algorithms are otherwise so similar, I feel a keyword argument is the way to go. But I have modified `sparseauxiva` to use the same API.